### PR TITLE
Enhancements to Maven Home

### PR DESCRIPTION
**Pull request by sarao1310**
- [x] All [tests](https://github.com/jfrog/build-info-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/build-info-go/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

File: build/maven.go
Comment (yahavi) [inline build/maven.go:209]: Missing checkerror

File: build/maven.go
Comment (yahavi) [inline build/maven.go:196]: We ignore the "err" if exist.
Let's add it to the created error.

File: build/maven_test.go
Comment (yahavi) [inline build/maven_test.go:82]:
```suggestion
	defer assert.NoError(t, mavenBuild.Clean())
```

File: build/maven_test.go
Comment (yahavi) [inline build/maven_test.go:122]: This requires table testing to remove code duplication. See how we do in TestGenerateBuildInfoForPython

I also find this guide useful:
https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

File: build/maven.go
Comment (yahavi) [inline build/maven.go:209]: Can we also return the absolute path for non-Windows systems?

**Review by yahavi — CHANGES_REQUESTED** [review]
Looks good @sarao1310.
Please consider my comments and please don't merge before my final approval.

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:189]: Done!

Comment (sarao1310) [reply build/maven.go:189]: it's been used two lines under in writing the debug message

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:209]: sure, but we will also need the if{} because we have to add ".cmd" when looking the path in windows

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:242]: Done!

File: build/maven_test.go
Comment (sarao1310) [reply build/maven_test.go:122]: Done

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:186]: done

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:242]: Done

File: build/maven_test.go
Comment (sarao1310) [reply build/maven_test.go:82]: I tried to change this, then the test is not passing any more

File: build/maven_test.go
Comment (sarao1310) [reply build/maven_test.go:122]: Done

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:209]: Done

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:196]: Done

File: build/maven.go
Comment (yahavi) [inline build/maven.go:186]:
```suggestion
```

File: build/maven.go
Comment (yahavi) [inline build/maven.go:187]: This line will panic if err is nil

File: build/maven.go
Comment (yahavi) [inline build/maven.go:190]: You return 2 similar outputs: output and stdout.
Let's remove the "output" and pass the other output to extractMavenPath.
Also, please keep in mind that both output and stdout are variable names without meaningful names.
```suggestion
		versionOutput, err := mm.execMavenVersion(maven)
```

File: build/maven.go
Comment (yahavi) [inline build/maven.go:197]: This line will panic if err is nil

File: build/maven_test.go
Comment (yahavi) [inline build/maven_test.go:110]: But you removed the whitespaces

File: build/maven_test.go
Comment (yahavi) [reply build/maven_test.go:83]: I disagree. This is exactly the same in that case.

File: build/maven_test.go
Comment (yahavi) [inline build/maven_test.go:106]: The "output" of what? Why appending?
Meaningful names are super important.

File: build/maven.go
Comment (yahavi) [inline build/maven.go:205]: Let's add a test for this new method

File: build/maven.go
Comment (yahavi) [inline build/maven.go:216]: Let's add a test for this new method

File: build/maven_test.go
Comment (yahavi) [inline build/maven_test.go:97]: Not meaningful names

**Review by yahavi — CHANGES_REQUESTED** [review]

**Comment by yahavi** [comment]
Please also rename the PR title to something that describes exactly what this PR does, such as - "Support Maven wrapper"

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:205]: This is just an if-else method, I don't think it needs a test

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:190]: Done

File: build/maven.go
Comment (yahavi) [reply build/maven.go:205]: That makes the test even easier :)

File: build/maven.go
Comment (sarao1310) [reply build/maven.go:216]: The result of this method is unexpected, because it is related to the operating system and the machine itself
this method executes the maven --version command, the version differs among different machines, also the maven home differs

File: build/maven_test.go
Comment (yahavi) [inline build/maven_test.go:144]: This unit test can be simplified by focusing on the target function. There is no need to read a Maven project and add a Maven module. For example, the following is enough for testing GetExecutableName when wrapper==false:
```go
	mavenModule := MavenModule{extractorDetails: &extractorDetails{}}
	mvnHome, err := mavenModule.getExecutableName()
	assert.NoError(t, err)
	assert.Equal(t, "mvn", mvnHome)
```

When the code is simple, you can easily cover more flows such as useWrapper==true:
```go
	mavenModule = MavenModule{extractorDetails: &extractorDetails{useWrapper: true}}
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Please fix the tests and the static analysis

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
This PR inclues multiple oprn discussions, so there's point for me to review it.

File: build/maven.go
Comment (eyalbe4) [inline build/maven.go:220]: Can be simplified to just - 
```
return
```

File: build/maven.go
Comment (eyalbe4) [inline build/maven.go:204]: The code in this function includes a lot of duplications can be significantly simplified.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I'd like to review this PR again following your changes.

File: build/maven.go
Comment (eyalbe4) [inline build/maven.go:190]: We should call the `determineError` function from inside the functions that use it, becuase it is responsible for creating the error the function returns.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: build/maven.go
Comment (eyalbe4) [inline build/maven.go:220]: The above description description has incorrect grammar - 
"For every different we show ..." 
"... that helps him to solve the error..."

Let's try rephrasing the description.

File: build/maven.go
Comment (eyalbe4) [inline build/maven.go:221]: determineError(mvnPath string, versionOutput string, err error) 
-->
determineError(mvnPath, versionOutput string, err error)

We should add CheckError to the errors returned by this function.

File: build/maven.go
Comment (eyalbe4) [reply build/maven.go:242]: @sarao1310 - see @yahavi's comment above. It looks like you resolved this conversation without add the CheckError.

File: build/maven.go
Comment (eyalbe4) [inline build/maven.go:243]: The above code - 
```
		maven = "./mvnw"
		if utils.IsWindows() {
			maven, err = filepath.Abs("mvnw.cmd")
		}
```
can be made more efficient as follows - 
```
		if utils.IsWindows() {
			maven, err = filepath.Abs("mvnw.cmd")
		} else {
			maven = "./mvnw"
		}
```

File: build/maven.go
Comment (eyalbe4) [reply build/maven.go:258]: You resolved this conversation without implementing my suggestion.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Pull request merged**
